### ISSUE TITLE
Experiment details implementation.

### ISSUE THREAD

**Issue by chenkang455** [wangst0181/pi-StepNFT#5]
@wangst0181 Hello, sorry to bother you again. During the process of reproducing this work, I noticed some fine-grained details and would like to ask for your advice.

1. I noticed that in the OpenPI model, the processing of $v$ only involves truncation of the chunk dimension, but no clipping is applied to the action-dim dimension: flow_v_snap[:, : self.config.action_chunk]. Is this an intentional design? In my own implementation, I found that adding clipping to action-dim makes the training less stable, which is a bit confusing for me.

2. I saw that the ablation studies provided a very comprehensive analysis, but I am a bit confused when looking at the code. For example, is the ODE implementation achieved by simply setting noise_level = 0? In my own ablation experiments, I found that if I directly convert SDE to ODE, the performance drops significantly. Upon analysis, I suspect the core difference is that my ODE implementation does not include the var term below. Besides, if noise_level = 0 here, then var = 1e-4, which directly causes the model gradients to explode. Could you please explain how the ODE is implemented in the code?
```
var = std_t_detached**2 + std_epsilon
E_pos = ((x0_pos - x0_target) ** 2 / var).sum(dim=dims)
E_neg = ((x0_neg - x0_target) ** 2 / var).sum(dim=dims)
```

3. The code supports training in $x_0$-space. However, my understanding is that in $x_0$-space, we cannot obtain v-old, meaning the $v$ sampled during rollout cannot be reused during training. I didn't see a corresponding v-old recomputation process in the code; could you point out where this is handled?

4. This question is regarding the NFT algorithm. NFT itself is off-policy; have you tried making it as "offline" as possible? For example, by collecting more data per rollout and performing more updates per iteration, similar to the RECAP process in PI0. Additionally, I noticed the EMA mechanism from Diffusion-NFT is implemented in the code, but I couldn't discern its advantages from the experiments. It seems that alpha = 0.1 is quite stable, and I have set it to alpha = 0.0 in my reproduction.

I would greatly appreciate your answers!

**Comment by wangst0181** [comment]
1. About the `flow_v_snap[:, :action_chunk]` line

There is no additional truncation on the action dimension in our implementation. At that point we only slice along the chunk dimension in order to match the training window.

You mentioned that explicitly truncating again along the action dimension leads to worse performance in your reproduction. That observation is not inconsistent with our code, because we are not doing that extra truncation. So I do not think this point should be interpreted as a missing step in the original implementation.

2. About the “original ODE / Diffusion-NFT-style” version

The original baseline is not obtained by simply setting `noise_level = 0` inside the current loss.

If one wants to reproduce the original version, two changes are required together:

* switch the sampler back to ODE, and
* change the loss back to the original form without the noise-aware correction.

This should be clear from the method design. In our current formulation, the step-wise objective is derived for the stochastic transition and its associated variance normalization. Therefore, directly zeroing the noise term inside the current loss is not equivalent to recovering the original formulation.

This is also exactly the distinction discussed in the paper: deterministic ODE rollouts, naive SDE rollouts, and SDE rollouts with noise-aware correction are different settings, not the same method with one scalar changed. Figure 2(a) and Section 5.3 make this separation explicit.

3. About “x0-space” and whether `v_old` can be reused

I think there is a terminology mismatch here.

In our implementation, changing the supervision to x-space / local-transition form does not mean that the model output semantics changed. The model still predicts the same velocity quantity as before. What changed is only how the loss is written.

So from our perspective, rollout-time `v_old` can still be reused. It is not being reinterpreted as a newly defined x0-space target. Rather, it remains the rollout-time velocity snapshot, and the loss is constructed around the sampled local transition using that same velocity space.

Therefore, the statement that “in x0-space we cannot obtain `v_old`, so the rollout velocity cannot be reused” does not apply to our implementation, because we did not change the model into an x0-prediction parameterization in the first place.

4. About the “more offline” setting

Our method is designed as an online method, not as a offline-style one.

This is important because the optimization data are supposed to be generated by the same rollout reference policy `v_old` that is used in the mirror construction. In other words, the sampled local transition and the reference velocity snapshot are intentionally matched within the same online rollout. This matching is part of the method design.

5. About EMA

EMA in our method is not introduced mainly as an optional trick for improving performance. Its role is more fundamental.

We use EMA for two purposes:

* to stay consistent with the original Diffusion-NFT-style setup, and
* to control the off-policy / on-policy gap so that `v_old` remains a meaningful rollout reference when constructing the mirrored branches.

The mirror construction in our method is centered on `v_old`:
`Δv = v_theta - v_old`,
`v_theta^+ = v_old + βΔv`,
`v_theta^- = v_old - βΔv`.

If `v_old` becomes identical to `v_theta`, then `Δv = 0`, and both mirrored branches collapse to the same point. In that case there is no branch separation, so the ranking signal disappears.

So the question is not whether EMA should provide an additional standalone gain by itself. The more important point is that some controlled lag between `v_old` and `v_theta` is necessary for the mirror construction to remain meaningful.

This is also consistent with the paper’s analysis and ablation: overly aggressive synchronization is unstable, while overly slow updates introduce off-policy lag. The decay schedule is there to balance these two effects, rather than to act as a cosmetic stabilization trick. See Algorithm 1, Section 4.1, and the decay analysis in Figure 4(c).

To summarize:

* we do not apply the extra action-dimension truncation you described;
* the original ODE-style baseline requires both ODE sampling and the original non-noise-aware loss, not merely `noise_level = 0`;
* the model output semantics were not changed, so reusing rollout-time `v_old` is valid in our implementation;
* the method is intentionally online, because the sampled local transition and the reference velocity snapshot are matched within the same rollout;
* EMA is used to maintain a proper reference policy for mirror construction, not simply as an optional performance trick.

**Comment by chenkang455** [comment]
@wangst0181 Thanks again for your detailed response — it was very helpful !

1. Regarding your first point, *“we are not doing that extra truncation”*: yes, that is correct. What I meant is that I tried adding an additional truncation along the action dimension on my side, but it introduced some training inconsistency.  Anyway, this is not a big problem.

2. Thanks for your clarification, but I was not able to find the corresponding ODE formulation. The following is the code I found in your repo:

```python
if use_x0_target:
    if x0_target is None:
        raise ValueError("use_x0_target=True requires `x0_target` to be provided.")
    if x0_target.shape != x_t.shape:
        raise ValueError(
            f"x0_target shape {x0_target.shape} must match x_t shape {x_t.shape}."
        )
    x0_pos = x_t - t_bc * v_pos
    x0_neg = x_t - t_bc * v_neg
    var = std_t_detached**2 + std_epsilon
    E_pos = ((x0_pos - x0_target) ** 2 / var).sum(dim=dims)
    E_neg = ((x0_neg - x0_target) ** 2 / var).sum(dim=dims)
    delta_E = E_pos - E_neg
```
My current understanding is that, in the ODE formulation, this part would simply become var = 1 instead of ``var = std_t_detached**2 + std_epsilon``. Is that the correct interpretation? I have tried this setting in my codebase but the performance keeps oscillating.

3. Thanks for clarifying the point about v_old. The reason I was confused is that I noticed the following code in your implementation:
``` python
explicit_inputs = kwargs.get("nft_inputs", None)
if explicit_inputs is not None:
    x_t = explicit_inputs["x_t"]
    t = explicit_inputs["timesteps"]
else:
    if "chains" not in data:
        raise ValueError("NFT forward requires `chains` or `nft_inputs`.")
    x_0 = data["chains"][:, -1].to(device)
    bsize = x_0.shape[0]
    t = torch.rand((bsize,), device=device)
    t_expanded = t[:, None, None]
    noise = torch.randn_like(x_0)
    x_t = (1 - t_expanded) * x_0 + t_expanded * noise
```
If execution goes into the else branch, then a new x_t is constructed. This seems consistent with Diffusion-NFT. However, in Diffusion-NFT, since this sample is not taken directly from the rollout process, we would normally need to recompute v_old. That is exactly where my confusion comes from.

Below is the code from diffusion-nft:
``` python
x0 = train_sample_batch["latents_clean"]

t = train_sample_batch["timesteps"][:, j_idx] / 1000.0

t_expanded = t.view(-1, *([1] * (len(x0.shape) - 1)))

noise = torch.randn_like(x0.float())

xt = (1 - t_expanded) * x0 + t_expanded * noise

with torch_autocast(enabled=enable_amp, dtype=mixed_precision_dtype):
    transformer_ddp.module.set_adapter("old")
    with torch.no_grad():
        # prediction v
        old_prediction = transformer_ddp(
            hidden_states=xt,
            timestep=train_sample_batch["timesteps"][:, j_idx],
            encoder_hidden_states=embeds,
            pooled_projections=pooled_embeds,
            return_dict=False,
        )[0].detach()
    transformer_ddp.module.set_adapter("default")

    # prediction v
    forward_prediction = transformer_ddp(
        hidden_states=xt,
        timestep=train_sample_batch["timesteps"][:, j_idx],
        encoder_hidden_states=embeds,
        pooled_projections=pooled_embeds,
        return_dict=False,
    )[0]
```

As for points 4 and 5, I have no further questions. Thanks again for your help!

**Comment by wangst0181** [comment]
To avoid further confusion, I will make the remaining points explicit here.

**2. About the ODE setting**

If you use the zero-noise case of the Flow-SDE transition, then at the solver level this is equivalent to using the Euler solver for ODE (which is suggested as shown in point 3). In that setting, the correct baseline is to use the `x0` target and to remove the variance normalization / noise-aware correction term. 

If you still observe collapse after switching to the Euler setting and changing the loss back to the original non-noise-aware form, then the issue is likely coming from some other configuration detail. In that case, you would need to provide the full relevant settings if you want me to judge whether the instability is caused by something else.

**3. About reusing `v_old`**

Only when
`collect_flow_snap = use_nft and solver_type_for_mode == "flow_sde"`
do we save the rollout-time inputs / snapshots.

So when using `euler`, the code path naturally falls back to the same construction logic as Diffusion-NFT. There is no need to confuse this with the Flow-SDE case.

For SDE, the rollout snapshot must be kept precisely because the supervision is defined on the actually noised trajectory transition sampled during rollout. For Euler / ODE, that specific requirement does not exist, so the construction reverts accordingly.

I will stop here for this thread. If you have any further questions, please contact me by email at `wangsiting2023@ia.ac.cn`, and include your WeChat ID in the email. I think it would be more efficient to continue detailed implementation discussions there rather than in a GitHub issue.

**Comment by chenkang455** [comment]
@wangst0181 Thanks for your efforts and detailed response! I have figured it out.
